### PR TITLE
Add arrow display styles to Vectors layer

### LINKED ISSUE #4149: Add arrow to vectors in napari vector layer
## 🚀 Feature
Add the feature that all vectors in a vector layer have arrows such that the direction they are pointing towards is clear.

## Motivation
It is very frustrating for me to debug my optical flow and related algorithms when I can not exactly see where the flow is pointing towards. You can see this in the image below.

image
image

Simple arrows would fix this and make my life a lot easier.

## Pitch

I want to have the option that every vector has an arrow at the tip such that the direction a vector is point to is clear.

### PR DESCRIPTION
**Pull request by jules-vanaret**
# Fixes/Closes
Closes #4149

# Description
This is a follow-up of the discussion #4149 in which adding different display style to the vectors of the Vectors layer was requested. As suggested by users @brisvag and @jni, I implemented a ComboBox in the GUI with 3 choices:

1. `rectangle` which is the old style of vectors, and currently the default value
2. `triangle` which displays vectors as triangles
3. `arrow` which displays a "classical" vector shape, with a rectangular shaft and a triangular head

napari_vectors_display_styles

Most of this contribution comes from additional functions to compute vectors meshes in _vispy/layers/vectors.py. Because changes are mainly in the function `generate_vector_meshes_2D`, there was no need to change anything in the functions `_on_data_change` or `generate_vector_meshes`, i.e it works natively with the 3D visualization.

The Vectors layer now provides a `string` parameter `edge_display_style`. Right now I check that any input string is one of `[rectangle, triangle, arrow]` in the setter of `edge_display_style` (in layers/vectors/vectors.py), but it does not feel like the right place. Also, I am not sure of my use of `trans.`.
I have not included any real tests, and I am pretty sure that I have forgotten to update the description of at least one function...

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
That looks great!
Thanks for the contribution!
I've activated CI and requested some reviews!
❤️

File: napari/_vispy/layers/vectors.py [hunk 1]
```diff
@@ -107,25 +110,127 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+                ]
+                if i % 3 == 1
+                else [
+                    7 * (i - 2) / 3 + 4,
+                    7 * (i - 2) / 3 + 5,
+                    7 * (i - 2) / 3 + 6,
+                ]
+                for i in range(3 * nvectors)
+            ]
+        ).astype(np.uint32)
```
Comment (brisvag) [inline napari/_vispy/layers/vectors.py]: black magic xD Is it possible to add some comments so this is easier to maintain for others in the future?

File: napari/layers/vectors/vectors.py [hunk 2]
```diff
@@ -436,6 +443,22 @@ def edge_width(self, edge_width: Union[int, float]):
...
+    @edge_display_style.setter
+    def edge_display_style(self, edge_display_style: str):
+        if edge_display_style in ['rectangle', 'triangle', 'arrow']:
+            # There should probably be a 'DisplayStyle' class
+            self._edge_display_style = edge_display_style
+        else:
+            raise
+
+        self.events.edge_display_style()
+        self.refresh()
```
Comment (brisvag) [inline napari/layers/vectors/vectors.py]: these should only be called if the style actually changes from before... though I see now that many other places don't do that :/

**Review by brisvag — COMMENTED** [review]
Awesome work! Love that you added the arrows as well, this is so much better than what we have :)

General suggestions:
- we should use an enum instead of strings to encode the different styles. You can look at how `Blending` works for examples on how to do that (base layer controls and base layer model).
	- I would also prefer `line` to `rectangle` which I find more intuitive
- I would just call it `display_style` or even just `style` or `mode` rather than `edge_display_style`... in fact I think we should ditch `edge` from everything in vectors, but that's another topic ^^' Surely @jni has opinions here

Other than that, I think having some more comments and docstrings in the new code would help reviewing, otherwise the math is a bit hard to follow :)

File: napari/layers/vectors/vectors.py (hunk 2 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py]: Do you mean that I should add a test like
```
old_style = self._edge_display_style

if old_style != edge_display_style:
    if edge_display_style in ['rectangle', 'triangle', 'arrow']:
         self._edge_display_style = edge_display_style
    else:
         raise
    self.events.edge_display_style()
    self.refresh()
```
I don't see any such test, for example for the blending mode. Have I misunderstood ?

File: napari/_vispy/layers/vectors.py (hunk 1 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Thank you for your positive review. 

- I am not (yet) well versed in all the typing conventions. I'll investigate `Blending` and make the corresponding changes.
- Please feel free to suggest better names for the styles ! "Rectangles" and "triangles" were chosen to be as objective as possible, but they definitely sound a little boring... ;)
- I also think that `style` would be simpler but wanted to stick to what was done in the rest of the code (I took inspiration from `edge_color_mode`). I'll rename to `style`.
- The functions to create the vertices and triangles are definitely under commented, but I wanted to "pre-push" the PR while I was thinking about the docstrings and comments. For the docstrings, I just discovered the one from the class `InteractionBoxHandle` in  `/layers/base/_base_constants.py` which is what I should aim for in terms of clarity of explanation for my choice of vertices.

File: napari/layers/vectors/vectors.py (hunk 2 above)
Comment (brisvag) [reply napari/layers/vectors/vectors.py]: Yeah, indeed this is missing for other places, but it shouldn't :P

I would do something like this (assuming a `VectorStyle` enum like I suggested above):

```py
    @edge_display_style.setter
    def edge_display_style(self, edge_display_style: str):
        old_style = self._edge_display_style
        self._edge_display_style = VectorStyle(edge_display_style) # this already takes care of checking and raising exceptions
        if self._edge_display_style != old_style:
            self.events.edge_display_style()
            self.refresh()
```

File: napari/_vispy/layers/vectors.py (hunk 1 above)
Comment (brisvag) [reply napari/_vispy/layers/vectors.py]: You did a great job at sticking to the style of the file, I just happen to disagree with it already, so that's on me :) And yeah, I always pre-push without comments at first just to make sure I'm on the right track. Now it's time to actually check the math to avoid broken edge cases, so comments are gonna be useful for reviewing.

> For the docstrings, I just discovered the one from the class InteractionBoxHandle in /layers/base/_base_constants.py which is what I should aim for in terms of clarity of explanation for my choice of vertices.

Sounds good!

File: napari/layers/vectors/vectors.py (hunk 2 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py]: I see, thanks for typing the whole stuff ! The modifications should (hopefully) arrive in the next 48h.

Comment (brisvag) [reply napari/layers/vectors/vectors.py]: no rush ;)

**Comment by brisvag** [comment]
Quick note so I don't forget: make sure that `display_out_of_slice` still works!

File: napari/_qt/layer_controls/qt_vectors_controls.py [hunk 3]
```diff
@@ -71,6 +76,23 @@ def __init__(self, layer) -> None:
...
+                vector_style_comboBox.setCurrentIndex(index)
+
+        self.vector_style_comboBox = vector_style_comboBox
+        self.vector_style_comboBox.currentTextChanged.connect(
+            self.change_vector_style
+        )
+
+        # I had added the line below to mimic what is done with
+        # color mode, but it leads to a crash. Is it necessary?
+        # self._on_vector_style_change()
```
Comment (brisvag) [inline napari/_qt/layer_controls/qt_vectors_controls.py]:
```suggestion
```

Not necessary, since you already initialized it correctly with `setCurrentIndex` above.

**Review by brisvag — COMMENTED** [review]
Awesome work! Thanks for the beatiful comments and docstrings, super valuable :) Sorry this slipped through the cracks, I hope you still have time to follow up. Don't hesitate to ping me directly if this happens!

One general issue: It seems that the `length` parameter interacts only with the position of the *tip* of the vector. This results in the `arrow` style being quite deformed by the length change. Try this code and change the length to 10. Even worse, if you go smaller than 1 the arrow folds onto itself xD

```py
import napari
import numpy as np
v = napari.Viewer(ndisplay=3)
a = np.array([[0, 0, 0], [10, 10, 10]])
v.add_vectors(a, vector_style='arrow')
```

Other than that, I think we're basically ready to merge!

File: napari/_vispy/layers/vectors.py [hunk 4]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+            else [
+                7 * (i - 2) / 3 + 4,
+                7 * (i - 2) / 3 + 5,
+                7 * (i - 2) / 3 + 6,
+            ]
+            for i in range(3 * nvectors)
         ]
     ).astype(np.uint32)
 
+    # np.tile([[0,1,2],[1,2,3],[4,5,6]], (nvectors, 1)) + np.repeat(7*np.arange(nvectors), 3)[:,None]
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]: This didn't work? I think you're very close! 😃

File: napari/_vispy/layers/vectors.py [hunk 5]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+        )
+
+    elif vector_style == 'arrow':
+        vertices, triangles = generate_meshes_arrow_2D(
+            vectors, width, length, p
+        )
+
+    else:
+        raise NotImplementedError
+
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]: This shouldn't be possible since we are using an enum?

File: napari/layers/vectors/vectors.py [hunk 6]
```diff
@@ -159,6 +172,7 @@ def __init__(
         properties=None,
         property_choices=None,
         edge_width=1,
+        vector_style='line',
```
Comment (jni) [inline napari/layers/vectors/vectors.py]: I know it's backwards incompatible, but I think the default should be 'triangle'.

File: napari/layers/vectors/vectors.py [hunk 7]
```diff
@@ -436,6 +453,27 @@ def edge_width(self, edge_width: Union[int, float]):
...
+            VectorStyle.ARROW
+                Displays vectors as arrows.
+        """
+        return str(self._vector_style)
+
+    @vector_style.setter
+    def vector_style(self, vector_style: str):
+        old_vector_style = self._vector_style
+        self._vector_style = VectorStyle(vector_style)
+        if self._vector_style != old_vector_style:
```
Comment (jni) [inline napari/layers/vectors/vectors.py:471]: It should be relatively straightforward to add a test for this:

- make a vectors layer
- connect the vector style event to some counter
- change the vector style to the same thing
- check counter hasn't changed
- change the vector style to a new thing
- check the counter has incremented

Does this make sense @jules-vanaret?

File: napari/layers/vectors/vectors.py [hunk 8]
```diff
@@ -588,7 +626,12 @@ def _view_face_color(self) -> np.ndarray:
         """(Mx4) np.ndarray : colors for the M in view vectors"""
         face_color = self.edge_color[self._view_indices]
         face_color[:, -1] *= self._view_alphas
-        face_color = np.repeat(face_color, 2, axis=0)
+
+        if self.vector_style == 'line':
+            face_color = np.repeat(face_color, 2, axis=0)
+
+        elif self.vector_style == 'arrow':
+            face_color = np.repeat(face_color, 3, axis=0)
```
Comment (jni) [inline napari/layers/vectors/vectors.py:632]: Please make a test that creates a vector layer and changes the style to each of the valid styles. That should cover many lines of code and make sure nothing explodes in the future 😂

File: napari/_vispy/layers/vectors.py [hunk 9]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+    triangles = np.arange(3 * nvectors).reshape((-1, 3)).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_arrow_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
+
+    Vectors are composed of 7 vertices and 3 triangles.
+    Vertices are generated according to the following scheme:
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]:
```suggestion
    Vertices are generated according to the following scheme:

```

File: napari/_vispy/layers/vectors.py [hunk 10]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+    vertices = vertices + width * offsets / 2
+
+    # faster than using the formula in the docstring
+    triangles = np.arange(3 * nvectors).reshape((-1, 3)).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_arrow_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]:
```suggestion
    """Generate mesh forming 2D arrows given input vectors.
```

Comment (jni) [reply napari/_vispy/layers/vectors.py]: Could you update the docstrings to conform with https://numpydoc.readthedocs.io/en/latest/format.html? Your descriptions are excellent but it would be great to have the argument list — I don't know what p does without delving into the code! 😊 

It's also unclear what length means for the vectors — isn't that implied in the data? I see that it's multiplied by the second component of the vector data but I don't understand why...

File: napari/_vispy/layers/vectors.py [hunk 11]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+
+    Vectors are composed of 3 vertices and 1 triangles.
+    Vertices are generated according to the following scheme:
+    1---x---0
+     .     .
+      .   .
+       . .
+        2
+
+
```
Comment (jni) [inline napari/_vispy/layers/vectors.py:230]:
```suggestion
```

File: napari/_vispy/layers/vectors.py [hunk 12]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+    ).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_triangle_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
+
+    Vectors are composed of 3 vertices and 1 triangles.
+    Vertices are generated according to the following scheme:
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]:
```suggestion
    Vertices are generated according to the following scheme:

```

File: napari/_vispy/layers/vectors.py [hunk 13]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
-            for i in range(len(vectors))
+            for i in range(2 * nvectors)
+        ]
+    ).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_triangle_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]:
```suggestion
    """Generate meshes forming 2D isosceles triangles to represent input vectors.
```

File: napari/_vispy/layers/vectors.py [hunk 14]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+        raise NotImplementedError
+
+    return vertices, triangles
+
+
+def generate_meshes_line_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
+
+    Vectors are composed of 4 vertices and 2 triangles.
+    Vertices are generated according to the following scheme:
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]:
```suggestion
    Vertices are generated according to the following scheme:

```

File: napari/_vispy/layers/vectors.py [hunk 15]
```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
...
+    1---x---0
+    | .     |
+    |   .   |
+    |     . |
+    3---v---2
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]: Possibly, for this to render correctly in sphinx, you need to indent by a further four spaces.

Comment (jni) [reply napari/_vispy/layers/vectors.py]: Yes, with a double-colon (`::`) above it, see:

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

Comment (jni) [reply napari/_vispy/layers/vectors.py]: Probably good to use this to preformat the numbers matrix below, too.

File: napari/_vispy/layers/vectors.py [hunk 16]
```diff
@@ -97,6 +100,8 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
         width of the line to be drawn
     length : float
         length multiplier of the line to be drawn
+    vector_style : str
+        display style of the vectors
```
Comment (jni) [inline napari/_vispy/layers/vectors.py]: This is not a string but a VectorStyle, right?

**Review by jni — CHANGES_REQUESTED** [review]
Thanks for making another push on this @jules-vanaret! I think it's very close! I've left some minor comments about docstring formatting and about testing. I think that it should be possible to add a test of the Qt components by adding a test to `napari/_qt/_tests/test_qt_viewer.py` that adds a vectors layer and then sets the style to various values. Hopefully I'm not wrong there, I have to say I find the Qt tests pretty intimidating myself 😅, but it's worth a shot! 🙏 Thank you! I'm really excited about getting this in!

File: napari/_vispy/layers/vectors.py (hunk 4 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Thank you for your feedback and the time you took to dive into the details of the code !
I have a busy week but I will try to implement your suggestions by Friday.

For this suggestion in particular, I found that this one-liner was probably faster, but definitely really hard to interpret at first glance. Do you think it would be a better idea to reimplement such lines (instead of the very broken down list comprehensions that are currently implemented) and to comment them profusely instead ?

File: napari/layers/vectors/vectors.py (hunk 6 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py]: I don't know your (the core developers) policy about insuring backward compatibility well enough to take a stand on this matter... :sweat_smile:
@brisvag do you have any preference ?

File: napari/layers/vectors/vectors.py (hunk 7 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py:471]: Yes that is clear ! Should the test be in `_vispy/_tests/test_vispy_vectors_layer.py` ?

File: napari/_vispy/layers/vectors.py (hunk 10 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py:250]: No problem. When I started working on this, only `generate_vector_meshes` was properly formatted; `generate_vector_meshes_2D` had only a single-line description, so I thought this was a choice.

Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Absolutely!
I just realized that many docstrings still had "Returns" fields with 
`vertices : (4N, 2) array for 2D and (8N, 2) array for 3D`
which I forgot to update... 

Would something like 
`vertices : (kN, 2) array for 2D and (2kN, 2) array for 3D, with k=4, 3, or 7 for vector_style='line', 'triangle', or 'arrow' respectively`
be appropriate ?

Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Regarding your comment about the `length` parameter: maybe it has been ill-named. The actual physical lengths of vectors are indeed directly given by the magnitude of the second component of the vector data, but the `length` parameter is an extra multiplicative factor that is applied to all vectors at the same time to allow the user to scale them in the viewer (e.g when the physical lengths are too small and are not directly visible on the viewer).

Maybe `scaling_factor` or `vector_scale` would be a better name ?

Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Last comment about the `length` parameter: right now, if the "length changed" event is triggered (same for the `width` parameter by the way), `generate_vector_meshes` is launched again so that every meshes and triangles are computed again, which is highly ineffective as only the positions of a few vertices would actually need to be multiplied by a scalar.

Maybe a next PR could tackle this optimization issue.

File: napari/_vispy/layers/vectors.py (hunk 4 above)
Comment (brisvag) [reply napari/_vispy/layers/vectors.py]: Definitely favour the speed in this part of the codebase. If it helps, you could leave a pseudocode version of the comprehension in a comment to explain what's going on.

File: napari/layers/vectors/vectors.py (hunk 6 above)
Comment (brisvag) [reply napari/layers/vectors/vectors.py]: I agree on the triangle default, and I don't think we shoudl worry about backward compatibility here. It only changes what it looks like, but any code that worked before will still work (unless it uses private stuff), and I can't think of cases where it will change so much as to become confusing/invisible, since the length is the same and it's now just tapering off.

File: napari/_vispy/layers/vectors.py (hunk 10 above)
Comment (brisvag) [reply napari/_vispy/layers/vectors.py]: Yeah, I always doung `length` was a bad name, but it's pretty central and user-facing, so if we change it we should introduce deprecations (shouln't be too hard to do, happy to help here if you need it). I like `scaling_factor`, or ever more explicitly `length_multiplier` or `length_scaling`. @jni?

File: napari/_vispy/layers/vectors.py (hunk 4 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: I reverted back to my previous implementation and tried to comment the process. I'll resolve the suggestion if you feel like it is enough. :wink:

File: napari/_vispy/layers/vectors.py [hunk 17]
```diff
@@ -377,25 +381,20 @@ def generate_meshes_arrow_2D(vectors, width, length, p):
...
+    #  [4,5,6]]
+    # as described in the docstring
+    vertices_pattern = np.tile(
+        [[0, 1, 2], [1, 2, 3], [4, 5, 6]], (nvectors, 1)
+    )
+    # 2. Add an offset to differentiate between vectors
+    triangles = (
+        vertices_pattern + np.repeat(4 * np.arange(nvectors), 2)[:, None]
+    )
+
```
Comment (brisvag) [inline napari/_vispy/layers/vectors.py]: oh wow, I didn't even know about slicing with `None`! Pretty cool :)

File: napari/_vispy/layers/vectors.py (hunk 4 above)
Comment (brisvag) [reply napari/_vispy/layers/vectors.py]: Actually, I find the numpy code more readable as well, so I think this is good enough!

File: napari/layers/vectors/vectors.py (hunk 7 above)
Comment (brisvag) [reply napari/layers/vectors/vectors.py:471]: Yes!

File: napari/_vispy/layers/vectors.py (hunk 17 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Some people consider this bad practice, as `np.newaxis` should be used instead of `None`... Don't hesitate if you would  like me to change it. :)

Comment (brisvag) [reply napari/_vispy/layers/vectors.py]: I guess it's probably wise since it's more well known :)

File: napari/layers/vectors/vectors.py (hunk 7 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py:471]: I just commited a version of such a test.

File: napari/layers/vectors/vectors.py (hunk 8 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py:632]: I am trying to implement a test in `napari/_qt/_tests/test_qt_viewer.py` as you suggested in your original message.
Would something like that be enough ?
```python
@pytest.mark.parametrize('dim', [2, 3])
def test_vector_style_change(make_napari_viewer, dim):
    viewer = make_napari_viewer(ndisplay=dim)

    vector_data = np.array([[[0] * dim, [1] * dim]])
    vector_layer = viewer.add_vectors(vector_data)

    vector_layer.vector_style = 'line'
    vector_layer.vector_style = 'triangle'
    vector_layer.vector_style = 'arrow'
```

File: napari/_vispy/layers/vectors.py (hunk 10 above)
Comment (jules-vanaret) [reply napari/_vispy/layers/vectors.py]: Thanks @brisvag, this could be another interesting project for a PR. If you have any references in the current napari codebase in which a parameter's name is being replaced (and deprecation notices are implemented), I could start taking a look at it.

File: napari/layers/vectors/vectors.py (hunk 8 above)
Comment (jules-vanaret) [reply napari/layers/vectors/vectors.py:632]: @brisvag Do you know if the test above covers the situation described by @jni above ?

Comment (jni) [reply napari/layers/vectors/vectors.py:632]: I think that works @jules-vanaret, sorry for the silence. I'm investigating the test failures now...

Comment (jni) [reply napari/layers/vectors/vectors.py:632]: LOL It *was* enough, but instead of using `make_napari_viewer` you used `viewer = napari.Viewer()` in your test and that was the failure. 😅 See d65a030. @Czaki it's only because both CI and local tests were failing in the same weird way that I noticed this — I wonder whether we could have a CI or pre-commit check that ensures that `Viewer` doesn't get directly instantiated in tests?

**Review by jni — APPROVED** [review]
assuming d65a030 fixes the test suite, this is ready imho! 🚀

**Comment by jni** [comment]
WHOO ALL GREEN! 🚀

**Comment by jules-vanaret** [comment]
Oops, I was on vacation so I missed the conclusion of the PR ! :sweat_smile: 
Many thanks to @jni and @brisvag for your patience and guidance. That was my very first PR to the Napari project, and I learnt a great deal. Hopefully more PRs will follow in the near future. :wink:  

For the sake of completeness regarding the Vectors layer, I have a few suggestions for follow-up PRs:

1. The PR was merged before I could finish tidying up all the comments and docstrings. This is minor, but I will try to find the time to finish it in a follow-up mini-PR.

> Yeah, I always doung `length` was a bad name, but it's pretty central and user-facing, so if we change it we should introduce deprecations (shouln't be too hard to do, happy to help here if you need it). I like `scaling_factor`, or ever more explicitly `length_multiplier` or `length_scaling`. @jni?
2. In the message above, @brisvag  suggested renaming `length` for something more suggestive, e.g `length_multiplier`.  If anyone versed in deprecations could help me on this one, this could be a quick fix.


> Last comment about the `length` parameter: right now, if the "length changed" event is triggered (same for the `width` parameter by the way), `generate_vector_meshes` is launched again so that every meshes and triangles are computed again, which is highly ineffective as only the positions of a few vertices would actually need to be multiplied by a scalar.
> 
> Maybe a next PR could tackle this optimization issue.
3. In the message above, I point out an optimization issue regarding the `length` and `width` parameters: when they are changed, vectors are recomputed from scratch but only a few vertices/meshes would actually need to be modified.

4. Right now vectors are always specified and drawn starting from their base (the "tail" of the vector). A combobox could be added to choose the "tethering point" of the vectors. To me, offering this choice to the user is important because (i) the current implementation is useful when you want to draw simple vectors, (ii) tethering at the center of the vector could be useful for displaying nematics fields (in a nutshell, vectors without direction e.g liquid crystals), (iii) tethering at the tip of the vector could be useful to easily make arrows that point directly to a given position.

If any of these ideas are of particular interest to the community, please let me know!

**Comment by brisvag** [comment]
All good options for future PRs!

> 1. The PR was merged before I could finish tidying up all the comments and docstrings

Always good to keep those updated, if we have missed something in the review :sweat_smile: 

The rest are all good options for followup PRs. I wouldn't call any of them high priority (for myself or from what I've seen discussed in the past), so feel free to tackle whichever is most blocking for your own work, and I'll be happy to review!

**Comment by jni** [comment]
> In the message above, @brisvag suggested renaming length for something more suggestive, e.g length_multiplier. If anyone versed in deprecations could help me on this one, this could be a quick fix.

@jules-vanaret currently, this change is unreleased, so a quick followup PR would *not* need to go through the deprecation pathway!

Currently I don't have this slated to go into 0.4.18 (that is a special release off a branch that requires manual cherry-picking of changes to that branch), so we would have time, but I'm quite tempted to add it — if you think you could change that in the next ~24h I could add both and there would be no need for a deprecation!

Having said that 0.5 will contain *lots* of deprecations (that's why we are making 0.4.18), so one more doesn't bother me. 😅 

> tidying up the comments and docstrings

Definitely not urgent, but a PR would definitely be appreciated! 😊 

> optimization issue regarding the length and width parameters

As above — performance fixes (especially for relatively "rare" events) can be made in subsequent PRs — and will be much appreciated. 😊 But they are not essential.

> Right now vectors are always specified and drawn starting from their base (the "tail" of the vector). A combobox could be added to choose the "tethering point" of the vectors.

I like this!

For all of the issues in that list @jules-vanaret, if you are not fixing them more or less immediately, it would be good to create separate issues so we keep track of them! 🙏

**Comment by brisvag** [comment]
> @jules-vanaret currently, this change is unreleased, so a quick followup PR would _not_ need to go through the deprecation pathway!

No, we're talking about renaming the `Vectors.length` parameter, which has been there for years, so that definitely needs deprecation!

**Comment by jules-vanaret** [comment]
Thank you both for the feedback. I have tons of things to do this week so I probably won't be able to work on anything immediately. I'll try to create the new issues rapidly though.

**Comment by jni** [comment]
> No, we're talking about renaming the Vectors.length parameter, which has been there for years, so that definitely needs deprecation!

Ah, sorry, I get it now! But on the plus side this wasn't changed here, right?
